include/observable.h

```diff
@@ -0,0 +1,271 @@
+/*************************************************************************
+ *
+ * Project:  OpenCPN
+ * Purpose: General observable implementation with several specializations.
+ *
+ * Copyright (C) 2022 Alec Leamas
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,  USA.
+ **************************************************************************/
+
+#ifndef OBSERVABLE_H
+#define OBSERVABLE_H
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include <wx/config.h>
+#include <wx/event.h>
+
+/** Return address as printable string. */
+std::string ptr_key(const void* ptr);
+
+
+class ObservedVar;
+class ObservedVarListener;
+
+
+/**
+ *  Private helper class. Basically a singleton map of listeners where
+ *  singletons are managed by key, one for each key value.
+ */
+class SingletonVar {
+
+private:
+  static SingletonVar* getInstance(const std::string& key);
+  std::unordered_map<wxEvtHandler*, wxEventType> listeners;
```

Comment:
This whole class could just be a member of ObservedVar of type `std::unordered_map<std::string, std::unordered_map<wxEvtHandler*, wxEventType>>`

Reply:
Yes, could. But is it better? As it is, it's IMHO short with a clean interface.

Reply:
After a night's sleep: Inheriting std__vector  instead of a simple class would expand this considerably, adding all sorts of methods to the interface. While possible., it IMHO becomes worse.